ls1intum/Artemis
# `Communication`: Allow to filter posts for multiple contexts

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Closes #6199 

### Description

The posts filter within the communication tab now allows the user to select multiple contexts simultaneously. The other filter options weren't changed. 

### Steps for Testing

Prerequisites:
- 2 Users
- 2 Courses with enabled communication
- 2 Lectures and exercises within each course
- A few of course-wide/lecture/exercise posts

1. Log in to Artemis as one user
2. Navigate to Course Communication
3. Check if the filter works as expected
4. Log in as the other user simultaneously 
5. Create a new post and check if the list of posts gets updated for the other user correctly (i.e., iff the other user has selected the context of the new post within the filter)
6. Select same/different lectures/exercises with the users and check if the correct posts are shown. Check if posts are updated automatically if the other user sends a new post and has selected the same lecture/exercise, but not if a different lecture/exercise is selected

### Review Progress






#### Code Review
- [ ] Review 4
#### Manual Tests

### Test Coverage




#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [course-discussion.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-discussion/course-discussion.component.ts.html) | 96.73% | ✅ |
| [discussion-section.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Client-Tests/app/overview/discussion-section/discussion-section.component.ts.html) | 85.86% | ✅ |
| [metis.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis.service.ts.html) | 90.41% | ✅ |
| [metis.util.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis.util.ts.html) | N/A% | ❌ |
| [post.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/post.service.ts.html) | 96.36% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [PostRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.metis/PostRepository.html) | 92% | ✅ |
| [PostSpecs.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.specs/PostSpecs.html) | 99% | ✅ |
| [PostService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis/PostService.html) | 90% | ✅ |
| [PostContextFilter.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4750-8/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.dto/PostContextFilter.html) | 100% | ✅ |


### Screenshots


Before: Only one context can be selected
<img width="1323" alt="before" src="https://user-images.githubusercontent.com/44754405/234823198-cfd86cac-0d3b-45fd-b303-f64121a1dd19.png">

After: Multiple contexts can be selected in parallel
<img width="1323" alt="after" src="https://user-images.githubusercontent.com/44754405/234823288-fe757012-2021-4fbf-9409-00176543bf97.png">
